### ISSUE TITLE
HID Service

### ISSUE THREAD

**Issue by AGlass0fMilk** [ARMmbed/mbed-os-experimental-ble-services#5]
### Proposed UUID(s) or link to specification document:
https://www.bluetooth.org/docman/handlers/downloaddoc.ashx?doc_id=245140

## Overview:

The HID Service exposes data and associated formatting for HID Devices and HID Hosts.

The HID service can be used to create devices that are immediately compatible with desktop operating systems to achieve:

- Transfer of peripheral input data, such as mouse and keyboards
- Transfer of generic data in a way that is accessible through portable HID libraries (like python). This lets the operating system do the abstraction of BLE to HID device and allows developers to more easily create cross-platform applications that collect data from BLE devices.

**Comment by pan-** [comment]
@jpbrucker Do you think we can borrow https://github.com/jpbrucker/BLE_HID for an implementation ?

**Comment by jpbrucker** [comment]
@pan- Of course, it is licensed apache-2 :) It was written for an old version of mbed, though. I don't currently have time to work on it but would still appreciate this code being shared.

**Comment by pan-** [comment]
Thanks @jpbrucker thanks for providing a status, your BLE_HID code was great and will a solid base to build upon.

**Comment by pea-pod** [comment]
I was surprised when this was not implemented fairly soon out of the box just for @AGlass0fMilk's reasons: everything knows how to talk HID. It would be such a boon to be able to do some nifty things through python on a PC without having to write custom services.

Another bullet point is that some Adafruit Circuit Python boards already do this. I know there are different aims slightly, but still.

Finally, USB already has most of the HID types. It might be nice to somehow incorporate the HID types from the mbed USB library. I mention this because it is entirely possible to have BLE HID but not USB HID and would require structuring.

I was going to try to make something for my children that connected to a controller I have (mainly PS3 or PS4) over BLE, but I am not nearly as adept, so this has set on the shelf. I got as far as getting my ST BLE shield working with the heart rate monitor and connect to the bluetooth monitor on my iPhone, but my BLE muscles are not nearly as strong as they need to be to even adapt @jpbrucker's work, although I've seen it for a while.
